File: 程序员面试宝典/2.4.分割链表.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def partition(self, head: ListNode, x: int) -> ListNode:
        small = ListNode(0)
        big = ListNode(0)

        p = small
        q = big

        while head:
            if head.val < x:
                p.next = head
                p = p.next
            else:
                q.next = head
                q = q.next

            head = head.next

        # 这一步最重要，要不然链表无法结束
        q.next = None
        p.next = big.next
        return small.next

File: 程序员面试金典/2.4.分割链表.cpp
class Solution {
public:
    ListNode* partition(ListNode* head, int x)
    {
        ListNode* p1 = new ListNode(0);
        ListNode* p2 = new ListNode(0);

        ListNode* q1 = p1;
        ListNode* q2 = p2;

        while (head) {
            if (head->val < x) {
                q1->next = head;
                q1 = q1->next;
            } else {
                q2->next = head;
                q2 = q2->next;
            }
            head = head->next;
        }

        q2->next = NULL;
        q1->next = p2->next;
        return p1->next;
    }
};